keps/sig-cluster-lifecycle/kubeadm/20191229-kubeadm-move-out-of-kk.md

```diff
@@ -0,0 +1,635 @@
+
+##### Bazel vs go build
+
+Bazel was investigated as an option for building kubeadm from the new
+repository but the benefits are unclear. BUILD files present a maintenance
+challenge and also the bazel WORKSPACE and gazelle files need to be kept
+in sync with k8s. On the other hand building using “go build” can take
+around ~20 sec on a fast machine with fast internet.
+
+The majority voted to not use bazel in the k/kubeadm repository.
+
+If caching is required a build container can be used.
+
+##### Syncing tags and branches between k/k and k/kubeadm
+
+A brief discussion of the #prow maintainers revealed the proposal that
+a post-submit job can be used for that instead of prow plugins.
+
+Supposedly such a job will trigger each time a new branch is created
+or if a tag is pushed. This means that the post-submit can trigger any bash
+script or application to get the latest list of branches and tags in k/k
+and apply the same to the kubeadm repository.
+
+Given kubeadm will potentially move within a single cycle this means that
+not all tags and branches have to be synced - see the options in “Versions
+in the support skew”. So potentially there must be a minimum version
+check in the post-submit job.
+
+Scenarios:
+
+###### k/k creates a new release-x.yy branch
+
+The post-submit job checks out master in k/kubeadm, creates a new branch
+release-x.yy and pushes the branch.
+
+###### k/k ‘s release-x.yy branch is fast-forwarded to master
+
+Currently this is a manual step performed by someone on the release team.
+It might be something that the k/kubeadm maintainers have to do manually
+on a daily basis. Automation is possible but is tricky and needs further
+discussion. Can be done as a nice-to-have in the future.
+
+###### k/release starts building kubeadm version x.yy.
+
+There is a chicken and egg problem where a tag in the k/kubeadm repository
+cannot be added before the k8s release is cut. This means that the build
+tools in k/release must not depend on looking for a tag in the k/kubeadm
+repository when building kubeadm.
+
+The k/release tooling pulls out the k/kubeadm repo and checks out
+branch “release-x.yy” k/kubeadm build tooling must write a KUBE_VERSION
+file with static versions, but a dynamic build date/time.
+Kubeadm’s version information in the binary should be based on the
+KUBE_VERSION file.
+
+At this point the k/release tooling must tag the latest commit of this branch
+with a placeholder tag `pre-<version>`.
+
+###### k/k pushes a new tag
+
+The post-submit picks the change and based on the k/k tag version,
+checks out the according branch in k/kubeadm, finds the latest commit with
+`pre-<version>` tag and removes the tag. Then pushes the same `<version>` tag
+without the `pre-` prefix.
+
+This process needs to handle multiple tags present on the same commit.
+If there are multiple `pre-<version>` tag the post-submit job must only
+operate on the `<version>` that that matched the k/k tag.
```

Comment:
I'm not following while this would be required.  What prevents the k/kubeadm owners from, ahead of k/k, declaring commitid N shall be the commit used when building k/kubeadm at vX.YY.Z to match k/k's vX.YY.Z?  If k/kubeadm is independent of k/k, we should be able to tag and build each independently "when they are ready".

Reply:
noticed a typo in the text - `that that` -> `tag that`

> What prevents the k/kubeadm owners from, ahead of k/k, declaring commitid N shall be the commit used when building k/kubeadm at vX.YY.Z to match k/k's vX.YY.Z

manually following 4 releases at all times will not be that easy. the proposal is to automate tagging between k/k and k/kubeadm using k/release tooling and a post-submit job.

> I'm not following while this would be required. What prevents the k/kubeadm owners from, ahead of k/k,

in terms of multiple-tags on the same commit for the post-submit job this might not be a problem, but think it will be a problem for:
- building for CI
- building in general from k/kubeadm's GIT

i do not know if the bash version library in k/k handles this fine, but i think it does not.
this can happen if a branch is stale and has not seen any recent backports.

if a commit has multiple tags, what tag will it pick to add to the KUBE_VERSION file?

Reply:
A human could commit the next version string to the KUBE_VERSION file, at the point they are declaring kubeadm ready.  Or automation that is about to tag some $ver could insert that into a commit against the KUBE_VERSION file, when kubeadm is released following some other trigger for readiness.

Reply:
The main culprit seems to be 'git describe'.  We can avoid the issue by not doing that, or at least not without additional logic that's maybe something like:

```
git tag --contains $(git describe) | sort | head -n
```
with a "sort" function that does precisely what we know we want relative to our tagging scheme.

Reply:
i'm going to update the KEP to remove the pre-tag proposal.